File: 060_Copy List with Random Pointer.cpp
/*
Copy List with Random Pointer

Create a deep copy of a linked list where each node has a next and random pointer.
Approach: weave copied nodes, assign randoms, then unweave.
*/

#include <bits/stdc++.h>
using namespace std;

struct Node {
    int val;
    Node* next;
    Node* random;
    Node(int x): val(x), next(nullptr), random(nullptr) {}
};

Node* copyRandomList(Node* head) {
    if (!head) return nullptr;
    // weave
    Node* cur = head;
    while (cur) {
        Node* copy = new Node(cur->val);
        copy->next = cur->next;
        cur->next = copy;
        cur = copy->next;
    }
    // assign randoms
    cur = head;
    while (cur) {
        if (cur->random) cur->next->random = cur->random->next;
        cur = cur->next->next;
    }
    // unweave
    cur = head;
    Node* dummy = new Node(0);
    Node* copyTail = dummy;
    while (cur) {
        Node* copy = cur->next;
        cur->next = copy->next;
        copyTail->next = copy;
        copyTail = copy;
        cur = cur->next;
    }
    return dummy->next;
}

void printList(Node* head) {
    Node* cur = head;
    while (cur) {
        cout << "Val:" << cur->val;
        if (cur->random) cout << " Rand:" << cur->random->val;
        cout << " | ";
        cur = cur->next;
    }
    cout << endl;
}

int main() {
    Node* n1 = new Node(7);
    Node* n2 = new Node(13);
    Node* n3 = new Node(11);
    n1->next = n2; n2->next = n3;
    n1->random = nullptr; n2->random = n1; n3->random = n1;
    cout << "Original: "; printList(n1);
    Node* cpy = copyRandomList(n1);
    cout << "Copied:   "; printList(cpy);
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1) extra (excluding output)

File: 060_Copy List with Random Pointer.py
"""
Copy List with Random Pointer

Create a deep copy of a linked list where each node has a next and random pointer.
Approach: use dictionary mapping original -> copy.
"""

class Node:
    def __init__(self, x):
        self.val = x
        self.next = None
        self.random = None

def copy_random_list(head: 'Node') -> 'Node':
    if not head: return None
    old_to_new = {}
    cur = head
    while cur:
        old_to_new[cur] = Node(cur.val)
        cur = cur.next
    cur = head
    while cur:
        copy = old_to_new[cur]
        copy.next = old_to_new.get(cur.next)
        copy.random = old_to_new.get(cur.random)
        cur = cur.next
    return old_to_new[head]

def print_list(head: 'Node'):
    cur = head
    out = []
    while cur:
        r = cur.random.val if cur.random else None
        out.append(f"({cur.val}, rand={r})")
        cur = cur.next
    print(' -> '.join(out))


if __name__ == '__main__':
    n1 = Node(7); n2 = Node(13); n3 = Node(11)
    n1.next = n2; n2.next = n3
    n1.random = None; n2.random = n1; n3.random = n1
    print('Original:', end=' '); print_list(n1)
    c = copy_random_list(n1)
    print('Copied:  ', end=' '); print_list(c)

# Time Complexity: O(n)
# Space Complexity: O(n)
